jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +188,434 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
+  convergence is often slow for nearly symmetric operators.
+
+  Parameters
+  ----------
+  A: function
+     Function that calculates the linear map (matrix-vector product)
+     ``Ax`` when called like ``A(x)``. ``A`` must return array(s) with the same
+     structure and shape as its argument.
+  b : array or tree of arrays
+      Right hand side of the linear system representing a single vector. Can be
+      stored as an array or Python container of array(s) with any shape.
+
+  Returns
+  -------
+  x : array or tree of arrays
+      The converged solution. Has the same structure as ``b``.
+  info : None
+      Placeholder for convergence information. In the future, JAX will report
+      the number of iterations when convergence is not achieved, like SciPy.
+
+  Other Parameters
+  ----------------
+  x0 : array, optional
+       Starting guess for the solution. Must have the same structure as ``b``.
+       If this is unspecified, zeroes are used.
+  tol, atol : float, optional
+      Tolerances for convergence, ``norm(residual) <= max(tol*norm(b), atol)``.
+      We do not implement SciPy's "legacy" behavior, so JAX's tolerance will
+      differ from SciPy unless you explicitly pass ``atol`` to SciPy's ``gmres``.
+  restart : integer, optional
+      Size of the Krylov subspace ("number of iterations") built between
+      restarts. GMRES works by approximating the true solution x as its
+      projection into a Krylov space of this dimension - this parameter
+      therefore bounds the maximum accuracy achievable from any guess
+      solution. Larger values increase both number of iterations and iteration
+      cost, but may be necessary for convergence. The algorithm terminates
+      early if convergence is achieved before the full subspace is built.
+      Default is 20.
+  maxiter : integer
+      Maximum number of times to rebuild the size-``restart`` Krylov space
+      starting from the solution found at the last iteration. If GMRES
+      halts or is very slow, decreasing this parameter may help.
+      Default is infinite.
+  M : function
+      Preconditioner for A.  The preconditioner should approximate the
+      inverse of A.  Effective preconditioning dramatically improves the
+      rate of convergence, which implies that fewer iterations are needed
+      to reach a given error tolerance.
+  qr_mode : bool
+      If True, the algorithm builds an internal Krylov subspace using a QR
+      based algorithm, which reduces overhead and improved stability. However,
+      it may degrade performance significantly on GPUs or TPUs, in which case
+      this flag should be set False.
+
+  See also
+  --------
+  scipy.sparse.linalg.gmres
+  jax.lax.custom_linear_solve
+  """
+
+  if x0 is None:
+    x0 = tree_map(jnp.zeros_like, b)
+  if M is None:
+    M = _identity
+
+  try:
+    size = sum(bi.size for bi in tree_leaves(b))
+  except AttributeError:
+    size = len(tree_leaves(b))
```

Comment:
If you use `  b, x0 = device_put((b, x0))` first (like in `cg`), then you won't need this fall-back condition.

Reply:
Fixed